components/org.wso2.carbon.identity.api.server.common/src/main/java/org/wso2/carbon/identity/api/server/common/ContextLoader.java

```diff
@@ -88,21 +102,36 @@ public static URI buildURIForBody(String endpoint) {
      */
     public static URI buildURIForHeader(String endpoint) {
 
-        String tenantQualifiedRelativePath =
-                String.format(TENANT_CONTEXT_PATH_COMPONENT, getTenantDomainFromContext()) + SERVER_API_PATH_COMPONENT;
-        String url = IdentityUtil.getEndpointURIPath(tenantQualifiedRelativePath + endpoint, false, true);
-
-        URI loc = URI.create(url);
-        if (!loc.isAbsolute()) {
-            Message currentMessage = PhaseInterceptorChain.getCurrentMessage();
-            if (currentMessage != null) {
-                UriInfo ui = new UriInfoImpl(currentMessage.getExchange().getInMessage(), null);
-                try {
-                    return new URI(ui.getBaseUri().getScheme(), ui.getBaseUri().getAuthority(), url, null, null);
-                } catch (URISyntaxException e) {
-                    LOG.error("Server encountered an error while building the location URL with scheme: " +
-                            ui.getBaseUri().getScheme() + ", authority: " + ui.getBaseUri().getAuthority() +
-                            ", url: " + url, e);
+        String url;
+        URI loc = null;
+
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+            try {
+                // Creating the context path from the endpoint which needs to be added to build the absolute public url.
+                String context = SERVER_API_PATH_COMPONENT + endpoint;
+                url = ServiceURLBuilder.create().addPath(context).build().getAbsolutePublicURL();
+                loc = URI.create(url);
+            } catch (URLBuilderException e) {
+                LOG.error("Server encountered an error while building the location URL in tenant qualified scheme.");
```

Comment:
Here also need to do the same as above suggested by @mefarazath. Throw a relevant expectation from here and handle it where this API is consumed.